File: Lab 1 - Logika pierwszego rzędu/ex2.py
from pyswip import Prolog
from helper import py2pl

prolog = Prolog()

# Fakty + Reguły
prolog.consult("ex2.pl")

# Zapytania
list1 = [1,2,3,4]
list2 = [5,5,6,6]

queries = [
    f'not_in_list(5, {py2pl(list1)})',
    f'not_in_list(5, {py2pl(list2)})',
    f'all_different({py2pl(list1)})',
    f'all_different({py2pl(list2)})',
    f'domain({py2pl(list1)}, 1, 4)',
    f'domain({py2pl(list2)}, 1, 4)',
]

for query in queries:
    print(query, bool(list(prolog.query(query))))

File: Lab 1 - Logika pierwszego rzędu/ex2.pl
not_in_list(_, []).
not_in_list(Elm, [Head|Tail]) :- Elm \= Head, not_in_list(Elm, Tail).

all_different([]).
all_different([Head|Tail]) :- not_in_list(Head, Tail), all_different(Tail).

domain([], _, _).
domain([Head|Tail], Min, Max) :- between(Min, Max, Head), domain(Tail, Min, Max).